content/blog/2025-12-16-allo-xls-backend.md

```diff
@@ -0,0 +1,218 @@
+
+### Systolic Array Generation (Method 1)
+
+XLS organizes computation as communicating processes based on Kahn process network semantics. A `proc` represents a stateful concurrent process with three components:
+- A `config` function declaring and wiring communication channels
+- An `init` function establishing initial state values  
+- A `next` function implementing a single state transition by consuming inputs, updating state, and producing outputs
+
+Rather than attempt general lowering from arbitrary Allo programs to explicit state machines (an intractably hard problem), we adopt a pattern-matching approach that specifically recognizes systolic array structures in MLIR and generates corresponding XLS grid implementations. We extracted metadata from the MLIR that drives a builder. The builder constructs DSLX proc ASTs for the PE and for the systolic grid that is then serialized.
+
+To see the generated XLS, see this link: https://github.com/Nikil-Shyamsunder/allo-xls-backend/tree/main/allo/backend/xls/examples/systolic
+
+### Meta-Systolic Arrays (Method 2)
+
+Beyond the basic pattern-matching approach for library-style systolic arrays, we developed a lowering strategy for systolic arrays expressed using Allo's metaprogramming constructs. However, the key challenge is that Allo's `meta_if` constructs are compiled away before we see them—by the time our compiler receives the MLIR, the conditional code generation has already happened.
+
+When Allo compiles a dataflow kernel like:
+```python
+@df.kernel(mapping=[P0, P1])
+def gemm(A, B, C):
+    i, j = df.get_pid()
+    with allo.meta_if(i == 0):
+        # Load B columns
+        ...
+    with allo.meta_elif(j == 0):
+        # Load A rows
+        ...
+    with allo.meta_else():
+        # Interior PE: MAC computation
+        ...
+```
+
+The resulting MLIR contains separate functions for each grid position: `gemm_0_0`, `gemm_0_1`, `gemm_1_0`, `gemm_1_1`, etc. Each function has different code depending on which `meta_if` branch was selected for that `(i, j)` coordinate at compile time. The challenge of this approach was that we had to reverse-engineer the spatial structure and PE types from this unrolled representation by analyzing the behavior of each generated function.
+
+To solve this, we utilized pattern matching using `SystolicDetector.is_metaif_systolic()` method by searching for multiple `gemm_i_j` functions, functions containing multiply-accumulate logic (interior PEs), and functions with only loads or stream operations (edge handlers). A meta-if systolic array is detected when there are ≥4 `gemm_i_j` functions and at least one contains MAC computation. The `MetaIfSystolicTranslator` analyzes each unrolled function to classify its role (MAC PEs, input loaders, and drain PEs). From the function names, we extract grid dimensions. We also extract the K loop bound from interior PE functions. Once we have this metadata, we plug this into the builder from Method 1.
```

Comment:
Ah, I see: it's because you want to do the pattern-recognition thing again, rather than directly translating the Allo PEs into XLS PEs. It would be good to clarify this up front.